File: lib_base/src/main/java/com/niuda/a3jidi/lib_base/base/base/BaseActivity.java
/*
 * Copyright 2016 jeasonlzy(廖子尧)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.niuda.a3jidi.lib_base.base.base;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.widget.ImageView;
import android.widget.Toast;

import com.blankj.utilcode.util.NetworkUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.bumptech.glide.Glide;
import com.niuda.a3jidi.lib_base.R;
import com.niuda.a3jidi.lib_base.base.utils.SystemBarUtils;

import butterknife.ButterKnife;
import pub.devrel.easypermissions.EasyPermissions;

/**
 * (￣▽￣)"
 *
 * @version V1.0 :
 * @author: Administrator
 * @date: 2017-11-11 14:07
 */
public abstract class BaseActivity extends AppCompatActivity{

    private static final String TAG = "BaseActivity";

    @SuppressWarnings("unchecked")
    public <T extends View> T findView(int id) {
        return (T) findViewById(id);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState, @Nullable PersistableBundle persistentState) {
        super.onCreate(savedInstanceState, persistentState);
        // SystemBarUtils.initSystemBarTint(this, translucentStatusBar(), setStatusBarColor());
        setContentView(initLayout());
        ButterKnife.bind(this);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);              //避免切换横竖屏
        AppManager.getAppManager().addActivity(this);                                   //将当前Activity添加到容器
        initView();
        initListener();
        initData();
        if(!NetworkUtils.isConnected()){
            ToastUtils.showShort("请检查网络是否连接");
        }
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:// 点击返回图标事件
                finish();
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    // 子类可以重写改变状态栏颜色
    protected int setStatusBarColor() {
        return SystemBarUtils.getColorPrimary(this);
    }

    // 子类可以重写决定是否使用透明状态栏
    protected boolean translucentStatusBar() {
        return false;
    }

    // 添加布局
    protected abstract int initLayout();

    // 初始化view
    protected abstract void initView();

    // 初始化监听器的代码写在这个方法中
    public abstract void initListener();

    // 初始数据的代码写在这个方法中，用于从服务器获取数据
    public abstract void initData();


    @Override
    protected void onDestroy() {
        super.onDestroy();
        AppManager.getAppManager().removeActivity(this);
    }

    // 通过Class跳转界面
    public void startActivity(Class<?> cls) {
        startActivity(cls, null);
    }

    // 通过Class跳转界面
    public void startActivityForResult(Class<?> cls, int requestCode) {
        startActivityForResult(cls, null, requestCode);
    }

    // 含有Bundle通过Class跳转界面
    public void startActivityForResult(Class<?> cls, Bundle bundle, int requestCode) {
        Intent intent = new Intent();
        intent.setClass(this, cls);
        if (bundle != null) {
            intent.putExtras(bundle);
        }
        startActivityForResult(intent, requestCode);
    }

    // 含有Bundle通过Class跳转界面
    public void startActivity(Class<?> cls, Bundle bundle) {
        Intent intent = new Intent();
        intent.setClass(this, cls);
        if (bundle != null) {
            intent.putExtras(bundle);
        }
        startActivity(intent);
    }

    // 初始化 Toolbar
    public void initToolBar(Toolbar toolbar, boolean homeAsUpEnabled, String title) {
        toolbar.setTitle(title);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(homeAsUpEnabled);
    }

    public void initToolBar(Toolbar toolbar, boolean homeAsUpEnabled, int resTitle) {
        initToolBar(toolbar, homeAsUpEnabled, getString(resTitle));
    }

    public void showToast(String msg) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    private ProgressDialog dialog;

    public void showLoading() {
        if (dialog != null && dialog.isShowing()) return;
        dialog = new ProgressDialog(this);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        dialog.setMessage("请求网络中...");
        dialog.show();
    }

    public void dismissLoading() {
        if (dialog != null && dialog.isShowing()) {
            dialog.dismiss();
        }
    }

    public void displayImage(String url, ImageView imageView) {
        Glide.with(getApplicationContext())//
                .load(url)//
                .error(R.mipmap.ic_launcher)//
                .into(imageView);
    }

}

File: lib_base/src/main/java/com/niuda/a3jidi/lib_base/base/base/BaseActivity.kt
package com.niuda.a3jidi.lib_base.base.base

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.TypedValue
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import com.gyf.barlibrary.ImmersionBar
import com.niuda.a3jidi.lib_base.R
import kotlinx.android.synthetic.main.layout_toolbar.*


abstract class BaseActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ImmersionBar.with(this)
                .statusBarColor(R.color.colorPrimary)
                .init()  //必须调用方可沉浸式
        setContentView(layoutResId())
        onCreated(savedInstanceState)
    }

    protected abstract fun layoutResId(): Int
    protected abstract fun onCreated(savedInstanceState: Bundle?)

    fun setPageTitle(pageTitle: String) {
        supportActionBar?.title = ""
        toolbar_title.text = pageTitle
    }

    fun initToolbar(isShowHomeUp: Boolean): BaseActivity {
        initToolbar(toolbar, title.toString(), isShowHomeUp)
        return this
    }

    fun initToolbar(toolbar: Toolbar?, title: String, isShowHomeUp: Boolean) {
        if (toolbar == null) return
        setSupportActionBar(toolbar)
        supportActionBar?.setHomeButtonEnabled(isShowHomeUp)
        supportActionBar?.setDisplayHomeAsUpEnabled(isShowHomeUp)
        supportActionBar?.setDisplayShowHomeEnabled(isShowHomeUp)
        val titleText:View? = toolbar.findViewById(R.id.toolbar_title)
        if(titleText!=null&&titleText is TextView){
            supportActionBar?.title = ""
            titleText.text = title
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        finish()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val fragments = supportFragmentManager.fragments
        if (fragments.isNotEmpty()) {
            fragments.forEach {
                it.onActivityResult(requestCode, resultCode, data)
            }
        }
    }

    fun getColorValue(resId: Int): Int {
        return if (Build.VERSION.SDK_INT >= 23) {
            super.getColor(resId)
        } else {
            resources.getColor(resId)
        }
    }

    fun dip2px(context: Context, dpValue: Float): Int =
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dpValue, resources.displayMetrics).toInt()

    fun px2dip(context: Context, pxValue: Float): Int =
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_PX, pxValue, resources.displayMetrics).toInt()

    override fun onDestroy() {
        super.onDestroy()
        ImmersionBar.with(this).destroy(); //必须调用该方法，防止内存泄漏
    }
}
